src/openms/source/CHEMISTRY/Tagger.cpp

```diff
@@ -0,0 +1,195 @@
+
+      // if aa is "L", then also add "I" as an alternative residue and extend the tag again
+      // this will add redundancy, (and redundant runtime) but we avoid dealing with J and ambigous matching to I and L later on
+      if (aa == 'L')
+      {
+        tag.pop_back();
+        tag.push_back('I');
+
+        if (tag.size() >= min_tag_length_)
+        {
+#pragma omp critical (tags_access)
+          tags.push_back(tag);
+        }
+        getTag_(tag, mzs, j, tags, charge);
+      }
+      tag.pop_back();  // remove last string
+      ++j;
+    }
+  }
+
+  Tagger::Tagger(size_t min_tag_length, double ppm, size_t max_tag_length, size_t min_charge, size_t max_charge, const StringList& fixed_mods, const StringList& var_mods)
+  {
+    ppm_ = ppm;
+    min_tag_length_ = min_tag_length;
+    max_tag_length_ = max_tag_length;
+    min_charge_ = min_charge;
+    max_charge_ = max_charge;
+
+    const std::set<const Residue*> aas = ResidueDB::getInstance()->getResidues("Natural19WithoutI");
+
+    for (const auto& r : aas)
+    {
+      const char letter = r->getOneLetterCode()[0];
+      const double mass = r->getMonoWeight(Residue::Internal);
+      mass2aa[mass] = letter;
+    }
+
+    // for fixed modifications, replace the unmodified residue with the modified one
+    for (const auto& mod : fixed_mods)
+    {
+      const ResidueModification* rm = ModificationsDB::getInstance()->getModification(mod);
+      Residue r = *(ResidueDB::getInstance()->getResidue(rm->getOrigin()));
+      r.setModification(rm->getId());
+
+      // remove the unmodified residue
+      // this requires searching the map by value, but this is only done once when the Tagger is initialized
+      for (std::map<double, char>::iterator it = mass2aa.begin(); it != mass2aa.end(); ++it)
+      {
+        if (it->second == rm->getOrigin())
+        {
+          mass2aa.erase(it);
+          break;
+        }
+      }
+      const char name = rm->getOrigin();
+      const double mass = r.getMonoWeight(Residue::Internal);
+      mass2aa[mass] = name;
+    }
+
+    // for variable modifications, add an additional instance of the residue with the modified mass to the list
+    for (const auto& mod : var_mods)
+    {
+      const ResidueModification* rm = ModificationsDB::getInstance()->getModification(mod);
+      Residue r = *(ResidueDB::getInstance()->getResidue(rm->getOrigin()));
+      r.setModification(rm->getId());
+      const char name = rm->getOrigin();
+      const double mass = r.getMonoWeight(Residue::Internal);
+      mass2aa[mass] = name;
+    }
+
+    min_gap_ = mass2aa.begin()->first - Math::ppmToMass(ppm, mass2aa.begin()->first);
+    max_gap_ = mass2aa.rbegin()->first + Math::ppmToMass(ppm, mass2aa.rbegin()->first);
+  }
+
+  void Tagger::getTag(const std::vector<double>& mzs, std::vector<std::string>& tags) const
+  {
+    // start peak
+    if (min_tag_length_ > mzs.size()) return; // avoid segfault
+
+#pragma omp parallel for schedule(guided)
```

Comment:
using `omp critical` within a `getTag_(...)` is somewhat unexpected (and not neccessary).
You can simply create a separate vector for each thread and join them later.
This is cleaner and potentially also faster (more scalable in any case).

```
#pragma omp parallel
{
  std::vector<std::string> tags_local;
  #pragma omp for schedule(guided)
    ....
    getTag_(tags_local, ...)
    ...
  #pragma omp critical (join_tags)
  {
    tags.insert(tags.end(), tags_local.begin(), tags_local.end());
  }
}
  

}
```